### ISSUE TITLE
benchmark GPs against sklearn, gpy, gpflow, and gpml

### ISSUE THREAD

**Issue by yorkerlin** [shogun-toolbox/shogun#3054]
This entrance task is about doing benchmarks against [sklearn v0.18](http://scikit-learn.org/dev/modules/gaussian_process.html), [gpy](https://github.com/SheffieldML/GPy), [gpflow](https://github.com/GPflow/GPflow), and [gpml](http://www.gaussianprocess.org/gpml/code/matlab/doc/) in terms of speed, memory usage, and accuracy.

Some tasks:
* Gaussian process (GP) regression 
* GP binary classification using Logistic likelihood

Note that for GP binary classification, there are many implemented inference algorithms (eg, EP, Laplace, and KL) in Shogun. 

Hint:
* For GP binary classification,sklearn (v0.18dev0) only has the Laplace implementation.  
* There is no sparse GP implementation in sklearn (v0.18dev0)

**Comment by yorkerlin** [comment]
This task is an entrance task of [the GP project](https://github.com/shogun-toolbox/shogun/wiki/GSoC_2016_project_large_gps)

**Comment by vigsterkr** [comment]
@yorkerlin sounds great! as there's already emerging some benchmarks on PCA, shouldn't we try to put this into one framework? i.e. keep the code somewhere and have the possibility to update the benchmarking results with the updated libraries?

**Comment by yorkerlin** [comment]
@vigsterkr 
Yes, cool. We can put this under one framework.

**Comment by vigsterkr** [comment]
@yorkerlin we should use/extend this framework:
https://github.com/zoq/benchmarks
i can setup a buildbot to generate the output...

you can see some of the results here:
http://www.mlpack.org/benchmark.html

the shogun that was used was version 3.2.0...

**Comment by juancamilog** [comment]
I'm interested in getting started to work on this one. @yorkerlin, do you agree that the framework at https://github.com/zoq/benchmarks should be used? I could initially try testing GP regression exact inference, just  to get a hang of what's needed to setup the benchmarks.

**Comment by yorkerlin** [comment]
Yes. Let's do benchmarks against sklearn, gpy, and GPflow under the
framework.

On Tuesday, March 8, 2016, Juan Camilo Gamboa Higuera <
notifications@github.com> wrote:

> I'm interested in getting started to work on this one. @yorkerlin
> <https://github.com/yorkerlin>, do you agree that the framework at
> https://github.com/zoq/benchmarks should be used? I could initially try
> testing GP regression exact inference, just to get a hang of what's needed
> to setup the benchmarks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/shogun-toolbox/shogun/issues/3054#issuecomment-194011100>
> .
>


-- 
best,

wu lin

**Comment by karlnapf** [comment]
@juancamilog I am super curious on the results of this, are there any already?

**Comment by rcurtin** [comment]
If you wrote some Python scripts for the benchmarking scripts feel free to submit them upstream and we can include it in the benchmarks repository. :)

**Comment by karlnapf** [comment]
Note

**Comment by yorkerlin** [comment]
@karlnapf 
I am working on setting an optimized python tool stack for benchmarks.
I will use openBLAS, numba, and joblib if possible.
I wonder whether our implementation can beat the optimized python implementation or not.

**Comment by yorkerlin** [comment]
@karlnapf @vigsterkr @lisitsyn  
It seems `llvm` compiler is better than `gcc` in terms of code optimization.
Do you think whether we should choose a right compiler?

**Comment by karlnapf** [comment]
@yorkerlin dont you think it might be more clever to put your ideas into the mlpack benchmarks?
For the GPs, if Shogun is faster than the gpflow and gpy, or even just scikit learn, that is the best start.

I think we will stay with gcc for now ... ;)

**Comment by yorkerlin** [comment]
@karlnapf 
I will focus on sklearn now. 
Too many things to do this week. :(

**Comment by yorkerlin** [comment]
@karlnapf 
Shogun's GP computes gradients wrt hyper-parameters for gradient-based model selection by default.

I think sklearn does not compute these gradients by default.
https://github.com/scikit-learn/scikit-learn/blob/56d625f/sklearn/gaussian_process/gpc.py#L664

**Comment by karlnapf** [comment]
We should compare a benchmark where we also learn the hyper-parameters using ML2.

**Comment by yorkerlin** [comment]
what is more, these gradients are for `logistic` likelihood `only`.

The third derivative is for `logistic` likelihood `only`.
https://github.com/scikit-learn/scikit-learn/blob/56d625f/sklearn/gaussian_process/gpc.py#L353

Shogun's GP can use 'student's T', 'probit' and 'logistic' likelihood.

**Comment by yorkerlin** [comment]
@karlnapf 
For `gradient`-based model selection, Shogun's GP uses the `Method of Moving Asymptotes` method in NLOPT by default.  
Sklearn uses `fmin_l_bfgs_b` for hyper-parameter searching by default.

**Comment by yorkerlin** [comment]
@karlnapf 
some results based on the `usps3v5` dataset

shogun:
```
using SingleLaplacianInferenceMethodWithLBFGS
cost 0.12 seconds at training
the negative_log_marginal_likelihood is 537.3986
cost 0.37 seconds at prediction
```
```
using SingleLaplacianInferenceMethod
cost 0.16 seconds at training
the negative_log_marginal_likelihood is 537.3986
cost 0.40 seconds at prediction
```

sklearn:
```
the negative_log_marginal_likelihood is-537.252898937 
cost 0.75 seconds at training
cost 0.40 seconds at prediction
cost 0.12 seconds at prediction
```

**Comment by yorkerlin** [comment]
I will check the `negative_log_marginal_likelihood` result from gpml.
There may be  some numerical issue in shogun or sklearn.

**Comment by yorkerlin** [comment]
For prediction, Shogun use MC samples IIRC.
MC samples may have to do with why shogun is slower than sklearn in prediction.

**Comment by yorkerlin** [comment]
I will post the script soon.

**Comment by karlnapf** [comment]
Great, looking forward to see this.
Such complicated frameworks as GPs might really allow Shogun to play its muscles ...

**Comment by karlnapf** [comment]
@Ialong might be interested in this too

**Comment by Ialong** [comment]
@karlnapf yep, definitely interested. Not sure I'll be able to get properly started on this before the deadline tomorrow but could do it over the weekend.

@yorkerlin @karlnapf should I use mlpack then?

**Comment by yorkerlin** [comment]
@karlnapf 
It seems `GPML` cannot run at octave 3.8.2.
I will pose the results based on `Matlab 2015b`.

**Comment by yorkerlin** [comment]
@karlnapf 

the script:
http://nbviewer.jupyter.org/gist/yorkerlin/b742cfe1669170ec0a6a
the data:
https://gist.github.com/yorkerlin/ee9d99d573dd0dc7d105

**Comment by karlnapf** [comment]
It would be good to add such plots to the benchmark system by mlpack. @yorkerlin don't you think it can be made work with octave?

**Comment by karlnapf** [comment]
Why is sklearn 10 times slower in training but faster in testing?

**Comment by yorkerlin** [comment]
@karlnapf 
in sklearn, 
`predict` function: MAP estimator is used
`predict_proba` function: deterministic integration is used. (https://github.com/scikit-learn/scikit-learn/blob/e5c366f/sklearn/gaussian_process/gpc.py#L290)

in shogun,
`Monte Carlo` method is used as suggested in `GPML`

**Comment by karlnapf** [comment]
Good points. Make sure the benchmarks give the same results. It is meaningless to compare things that produce different results. Also if possible the same algorithm should be used. BUT, at the end of the day, what matter is "user wants to do gp regression with exact inference, which toolbox is fastest"

**Comment by karlnapf** [comment]
About vectorization: this is done via using linalg
About multithread: Yes, definitely this should be done, using openmp if possible

**Comment by yorkerlin** [comment]
@karlnapf 
This is for classification :)
If we have a lot of outliers, `Students' t` (approximate inference) should be better than `Gaussian` (exact inference) likelihood.
For regression (exact inference), I leave it to @Ialong

**Comment by yorkerlin** [comment]
@karlnapf 
Another case is `Cox` likelihood for survival analysis.

**Comment by yorkerlin** [comment]
@karlnapf 
For beginners or black-box users, the statement is true. 
For advanced users, they should consider speed, accuracy, and interpretability. 

` what matter is "user wants to do gp regression with exact inference, which toolbox is fastest"  `
For simple regression, I agree with the statement.
 
For other cases, it is not  always true. 
For example, I will not use `sklearn`,  if I care about `p-values` and `adjusted R^2`.

**Comment by yorkerlin** [comment]
@karlnapf 
check out some result about `GPy` at  http://nbviewer.jupyter.org/gist/yorkerlin/b742cfe1669170ec0a6a

**Comment by karlnapf** [comment]
Cool that Shogun is faster! :) Some things to do

 * we need to put this systematically into the mlpack benchmarks .. ipython notebooks get lost so we forget about the results. Then we can collect these results and publish at some point
 * you are right that use cases can differ and should be carefully investigated on their own
 * larger problems would be interesting, where training takes a few more seconds

**Comment by yorkerlin** [comment]
@karlnapf 
take a look at this http://nbviewer.jupyter.org/gist/yorkerlin/b742cfe1669170ec0a6a

FITC and Var_DTC are included.

**Comment by yorkerlin** [comment]
To summary,
Shogun's GP is faster than sklearn, GPML, and GPy for most of methods.

However, for Var_DTC, GPy has a parallel implementation. 
What is more, Var_DTC in GPy is as fast  as in Shogun. We should improve the Var_DTC method in Shogun.

**Comment by vigsterkr** [comment]
@yorkerlin AWESOME++
we should convert this to a http://github.com/zoq/benchmarks/ task, that way we could re-test this anytime we want :)

**Comment by karlnapf** [comment]
Really nice!

I agree on integrating this into the benchmark repo. @ialong might be interested

**Comment by ialong** [comment]
I am definitely interested, perhaps priority should be on #3043 for me now. I will try to get to benchmarking as well

**Comment by karlnapf** [comment]
Yeah maybe do that one. Or both :) The benchmarking should be a simple matter of more or less copying @yorkerlin 's code to the benchmark system. There were lots of patches for that recently, so it is a no-brainer

**Comment by yorkerlin** [comment]
@karlnapf 
take a look at this
I borrow some codes from @youssef-emad 

https://github.com/yorkerlin/plots-for-GP/blob/master/Untitled1.ipynb

**Comment by karlnapf** [comment]
Nice notebook.
But my question again is: Why is this useful in particular for GPs. 
I would much rather just include these plots in the existing notebook, for the *interesting* cases.

**Comment by yorkerlin** [comment]
My point is
1. We should have a notebook about "black box" classifiers. I thinks we
have it now.
2. We have have a notebook about kernels for advanced users since kernel is
one of Shogun's key features. automatical kernel selection will be another
option.
3. We should have a notebook about inference methods for GP if the default
method is too slow. Users may want to know how to choose a right inference
method for large scale GP.

**Comment by karlnapf** [comment]
I see.

Feel free to create issues for 2/3 but name (or rename existing) them properly, and describe exactly what you want to do and what the focus should be. Can all be done as part of GSoC, but needs proper specifications ....
